src/rust/engine/process_execution/src/sandboxer.rs

```diff
@@ -0,0 +1,290 @@
+//
+// The Sandboxer solves the problem via a separate, dedicated process that writes executables into
+// sandboxes, and never itself spawns any subprocesses.
+//
+// The Sandboxer code consists of the following units:
+// - A server, running in a standalone process, that receives and satisifes file writing requests.
+// - A client that sends file writing requests to that server via a Unix domain socket.
+// - The SandboxerProcess struct, which encapsulates a server process and a client connection to it.
+// - The Sandboxer struct, which manages a SandboxerProcess, respawning it as needed.
+//
+// We use a Unix domain socket primarily because it's easy to pick a unique socket path per repo.
+// The lower latency is a bonus.
+
+fn ensure_socket_file_removed(socket_path: &Path) -> Result<(), String> {
+    let mut ret = fs::remove_file(socket_path);
+    if let Err(ref err) = ret {
+        if err.kind() == io::ErrorKind::NotFound {
+            ret = Ok(());
+        }
+    } else {
+        info!("Removed existing socket file {}", &socket_path.display());
+    }
+    ret.map_err(|e| e.to_string())
+}
+
+// The service that the spawned sandboxer process runs.
+// The process exits if anything goes wrong, since it's easy and cheap to respawn it.
+// It also exits if it's been idle for a while, so that orphaned processes don't pile up.
+#[derive(Debug)]
+pub struct SandboxerService {
+    socket_path: PathBuf,
+}
+
+impl SandboxerService {
+    const POLLING_INTERVAL: Duration = Duration::from_millis(500);
+    const ALLOWED_IDLE_TIME: Duration = Duration::from_secs(60 * 15);
+    const ALLOWED_IDLE_BEATS: usize =
+        Self::ALLOWED_IDLE_TIME.div_duration_f32(Self::POLLING_INTERVAL) as usize;
+
+    pub fn new(socket_path: PathBuf) -> Self {
+        Self { socket_path }
+    }
+
+    fn ensure_socket_file_removed(&mut self) -> Result<(), String> {
+        ensure_socket_file_removed(&self.socket_path)
+    }
+
+    pub async fn serve(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+        // `bind()` expects the socket file not to exist, so we ensure that.
+        self.ensure_socket_file_removed()?;
+        let listener = UnixListener::bind(&self.socket_path)?;
+        // `bind()` creates the socket file, so now we can poll it, in case the
+        // user deletes it.
+        // NB: We can't easily use the notify crate to watch changes to the socket file,
+        //  because on MacOS the FSEvent watches are path-based, not inode-based, and they
+        //  are batched. So we may get a spurious notification for the removal above, and not
+        //  know if we should act on it. We could use notify's PollWatcher, but it's easier to
+        // implement simple polling ourselves, without the overkill of the notify crate.
+        let socket_path = self.socket_path.clone();
+        let idle_beats = Arc::new(AtomicUsize::new(0));
+
+        let idle_beats_ref = idle_beats.clone();
+        thread::spawn(move || loop {
```

Comment:
Performance question: Would it be acceptable to just spin up a sandboxer for each sandbox (and have it exit immediately)?

Reply:
Interesting thought. Maybe, in which case it could just be `cp` I suppose. The overhead of shelling out to `cp` to copy an empty file is (according to hyperfine) well under 5 ms. So that is probably what we should start out with...

Reply:
Sounds like a decent minimal start to a viable solution.